### PR TITLE
 feat: implements multiple lint rules

### PR DESCRIPTION
**Pull request by claymcleod**
This pull request adds two new rules to `wdl-lint`.

- **Rule Name**: `DeprecatedPlaceholderOption`
- **Rule Name**: `RuntimeSectionKeys`

These rules are part of the [baseline lint rules](https://github.com/stjude-rust-labs/wdl/discussions/11). There were other changes that were introduced that were complementary to adding this lint, and all of those have been called out in my notes below.

The biggest changes were:

- I pulled out the `Version` from a private struct within the AST validation to a public struct (named `SupportedVersion` to avoid confusion with the `Version` AST node). This provides a unified way to parse a WDL version so that it can be used across the libraries (and, in particular, the lint rules).
- I now sort the diagnostics that are output by the tests to ensure the order is deterministic.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.

Rule specific checks:

- [x] You have added the rule as an entry within `RULES.md`.
- [x] You have added the rule to the `rules()` function in `wdl-lint/src/lib.rs`.
- [x] You have added a test case in `wdl-lint/tests/lints` that covers every
      possible diagnostic emitted for the rule within the file where the rule
      is implemented.
- [x] You have run `wdl-gauntlet --refresh` to ensure that there are no 
      unintended changes to the baseline configuration file (`Gauntlet.toml`).
- [x] You have run `wdl-gauntlet --refresh --arena` to ensure that all of the 
      rules added/removed are now reflected in the baseline configuration file 
      (`Arena.toml`).

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

**Comment by claymcleod** [comment]
Apologies for the duplicates PRs (#115 and #119)—GitHub and I had a moment. I'll apply my previous comments here.

File: .github/pull_request_template.md
Comment (claymcleod) [inline .github/pull_request_template.md:31]: This was added after discussion with @peterhuene as its a common misstep when introducing a new visitor method.

File: wdl-ast/src/lib.rs
Comment (claymcleod) [inline wdl-ast/src/lib.rs:64]: This is where I pulled the `Version` out of validation and called it `SupportedVersion`.

File: wdl-ast/src/lib.rs
Comment (claymcleod) [inline wdl-ast/src/lib.rs:58]: The `V1` enum is in here now available at the top level under `version`.

File: wdl-grammar/src/diagnostic.rs
Comment (claymcleod) [inline wdl-grammar/src/diagnostic.rs:87]: Importantly, this PR includes some unexpected-but-important changes regarding the ordering of diagnostics to make the output of the tests deterministic. This will change the order of many other unrelated tests.

File: wdl-lint/tests/lints.rs
Comment (claymcleod) [inline wdl-lint/tests/lints.rs:131]: This is where I added the deterministic sorting of errors.

File: wdl-lint/tests/lints.rs
Comment (claymcleod) [inline wdl-lint/tests/lints.rs:148]: And, subsequently, the sorting of lint warnings.

File: wdl-lint/tests/lints/command-mixed-line-cont/source.wdl
Comment (claymcleod) [inline wdl-lint/tests/lints/command-mixed-line-cont/source.wdl:1]: From here on down, it's mostly excepting the new rule and changes to the ordering of non-related tests. You can take a look at `runtime-keys-wdl-1.0` and `runtime-keys-wdl-1.1` for the most important new tests.

File: wdl-lint/RULES.md
Comment (claymcleod) [inline wdl-lint/RULES.md:10]: This is due to the Markdown All-in-One extension formatting from within VSCode. I suppose it's fine unless anyone has any issues—that makes it much easier to keep the spacing consistent (for me at least).

**Comment by claymcleod** [comment]
@peterhuene and @adthrasher, though this thread is on the now-closed PR, [this](https://github.com/stjude-rust-labs/wdl/pull/119#discussion_r1673140726) still needs an opinion from you.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [inline wdl-lint/src/rules/runtime_section_keys.rs:158]: As `inspect_v1_1_key` is called once per runtime item, is there a way we could avoid the `collect` here?

Such as:

```rust
let mut entries = V1_1_KEYS.iter().filter(|e| e.name() == key);
let entry = entries.next();
assert!(entries.next().is_none(), "more than one key matched; duplicate entry in static key map");
match entry {
...
}
```

Also, it seems like the key sets could instead be a `HashMap<&'static str, Key>` where `Key` would be:

```rust
enum Key {
    /// A key that is deprecated in favor of another key.
    /// The value is the replacement key.
    Deprecated(&'static str),
    /// A runtime key that is recommended to be included.
    Recommended,
    /// A runtime key that has a reserved meaning in the specification but which
    /// execution engines are _not_ required to support. These are also called
    /// "hints" in WDL parlance.
    Hint,
    /// A runtime key that has a reserved meaning in the specification and which
    /// execution engines are required to support (but don't necessarily have to
    /// be present in WDL documents).
    ReservedMandatory,
}
```

That way we don't need to iterate the collection at all.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [inline wdl-lint/src/rules/runtime_section_keys.rs:283]: This seems like a lot of allocations, a `clone` here, and then both a `clone` and `collect` in `recommended_keys`.

An alternative might be:

```rust
let empty = Default::default();
let recommended_keys = match minor_version {
    V1::Zero => &V1_0_KEYS,
    V1::One => &V1_1_KEYS,
    _ => &empty,
};

for key in recommended_keys.iter().filter(|key| key.is_recommended())  {
    ...
}
```

File: wdl-lint/tests/lints/deprecated-placeholder-options-v1.1/source.errors
Comment (peterhuene) [inline wdl-lint/tests/lints/deprecated-placeholder-options-v1.1/source.errors:5]: I think the span should probably be just the identifier of the node (e.g. `sep`), rather than the entire node.

We can probably add a `name(&self) -> Ident` method to each of the options to facilitate this.

**Review by peterhuene — COMMENTED** [review]
This looks great; a few comments to go over.

File: wdl-lint/tests/lints/deprecated-placeholder-options-v1.1/source.errors
Comment (peterhuene) [reply wdl-lint/tests/lints/deprecated-placeholder-options-v1.1/source.errors:5]: On second thought, given that the remediation for the diagnostic is to remove the entire option, having the highlight be the entire option makes sense to me. 

Resolving this comment.

File: Cargo.toml
Comment (claymcleod) [reply Cargo.toml:31]: Oh, I was not aware of that. Okay, thanks for the tip—I'll try that instead.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs:158]: So I originally had it the way you describe above and ended up flipping back and forth. There were some pros I liked in terms of the clarity of the code with the `HashSet` option (from the top of my head, duplicates keys would be able to be detected rather than re-inserting over an existing key + pulling out the `DeprecatedKey` made it possible to statically verify only deprecated keys could be passed to `deprecated_runtime_key` for future implementers).

All that said, I made those decisions because I felt the performance hit was small compared to a small bump up in clarity of the code but, as you and I discussed, we can take the philosophy of eeking out every bit of performance here.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs:283]: Yeah, this is a case of not coming back to fix this: there were some quirks of `lazy_static` that I was looking at trying to avoid this, but I see now that I got a working solution and never actually came back to it. With `std::sync::OnceLock`, this appears to be much more straightforward.

File: Arena.toml
Comment (claymcleod) [reply Arena.toml:1165]: Yeah I recall this quirk (maybe Cromwell uses `disk` but WDL settled on `disks`)? Should this special case mention Cromwell? I need some input from you and @a-frantz on the particulars around this case.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [reply wdl-lint/src/rules/runtime_section_keys.rs:158]: If detecting duplicates on init is desired, I think we could structure initialization to look like:

```rust
V1_1_KEYS.get_or_init(|| {
    const KEYS: &[(&str, Key)] = [("container", Key::ReservedMandatory), ...];

    let mut keys = HashMap::with_capacity(KEYS.len());
    for (k, v) in KEYS {
        let prev = keys.insert(k, v);
        assert!(prev.is_none(), "duplicate key `{k}`");
    }
    
    keys
});
```

File: Arena.toml
Comment (a-frantz) [reply Arena.toml:1165]: It's an archaic Cromwell quirk that has persisted in our repo and I'm not sure it it's present anywhere else. [The current state of affairs](https://cromwell.readthedocs.io/en/stable/RuntimeAttributes/#disk) is that even Cromwell has dropped support for `disk` (mostly). I don't remember offhand the exact history, but I'm not sure it matters at this point.

I'm about to look through Arena to see if this is present in any repo other than ours. I suspect it won't be. I'm not certain this needs any special handling. We just need to fix it in `workflows`, and then hopefully we won't see this "in the wild" anywhere else.

Comment (a-frantz) [reply Arena.toml:1165]: Yea that's what I thought. I expanded Arena to look through every Gauntlet repo, and then `grep`ed the resulting Arena.toml for this diagnostic. It only appears in `workflows`. I still think the current diagnostics emmitted by `RuntimeSectionKeys` needs to be consolidated, but no special handling for `disk`/`disks` is needed IMO.

Comment (claymcleod) [reply Arena.toml:1165]: Okay, that sounds good, I will roll with Andrew F.'s suggestion if that's good with you @adthrasher

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs:57]: It's a good question and I considered that briefly too. To me, since this will be hidden as a behind-the-scenes implementation detail combined with the fact that the specification refers to them as "reserved, mandatory keys", I felt it was okay to go with this.

**Comment by claymcleod** [comment]
I've now hammered out a draft of all of the changes that were requested on #120—there may be some sloppiness, such as unnecessary allocations that can be cleaned up, but I'm going to push it up now so others can review the overall output with the new implementation. I'll take a look for any optimizations tomorrow.

File: wdl-analysis/tests/analysis.rs
Comment (claymcleod) [inline wdl-analysis/tests/analysis.rs:113]: FWIW, this was a stray lint warning that clippy found and I just corrected here.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (adthrasher) [inline wdl-lint/src/rules/runtime_section_keys.rs:432]: What happens if someone does this?
```
task foo {
  .... 
  runtime {
    cromwell: {
      # cromwell specific 
      ...
    }
    miniwdl: {
      # miniwdl specific
      ...
    }
  }
}
```
That's an example in the spec (https://github.com/openwdl/wdl/blob/1.1.1/SPEC.md#conventions-and-best-practices).

Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs:433]: That's a good question. The specification is clear that any non-reserved keys are deprecated:

> 🗑 Additional arbitrary attributes may be specified in the runtime section, but these may be ignored by the execution engine. These non-standard attributes are called "hints". The use of hint attributes in the runtime section is deprecated; a later version of WDL will introduce a new hints section for arbitrary attributes and disallow non-standard attributes in the runtime section.

That said, the example does still exist in the specification. My feeling is that you should be able to `#@ except` those entries.

Comment (a-frantz) [reply wdl-lint/src/rules/runtime_section_keys.rs:433]: > That said, the example does still exist in the specification. My feeling is that you should be able to `#@ except` those entries.

Sounds fine to me. I ofc have not memorized the docs for all the execution engines out there, but I don't recall ever actually seeing anyone adopt this convention. I think it's a weird hypothetical that most likely doesn't belong in the spec in the first place.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [inline wdl-lint/src/rules/runtime_section_keys.rs:454]:
```suggestion
            //   one of the two recommended keys is omitted, and that is handled at the end
```

File: wdl-lint/src/rules/deprecated_placeholder_option.rs
Comment (peterhuene) [inline wdl-lint/src/rules/deprecated_placeholder_option.rs:69]: I'm not quite sure why this is `Option<Option<SupportedVersion>>`. Rules may expect that `document` is called before any other callbacks as it is the root of the walk, so should this not be `Option<SupportedVersion>`?

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [inline wdl-lint/src/rules/runtime_section_keys.rs:360]: The fix for the issue you were seeing:

```suggestion
                self.encountered_keys.clear();
                self.non_reserved_keys.clear();
```

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [inline wdl-lint/src/rules/runtime_section_keys.rs:428]: So I think this isn't quite correct, as we'll still process the items of the additional runtime sections.

I think we should have a flag, reset on task entry, which indicates that we've exited from a runtime section; if that flag is set, have future calls to `runtime_section` and `runtime_item` return early.

**Review by peterhuene — COMMENTED** [review]
Looks great, just four more additional comments.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs]: I ended up changing the language before I saw this comment to exactly what you say here.

File: wdl-lint/src/rules/deprecated_placeholder_option.rs
Comment (claymcleod) [reply wdl-lint/src/rules/deprecated_placeholder_option.rs:69]: You're correct that this assumption _should_ always remain true. The virtue that lead me to split into two `Option`s is simply to be able to handle the two different meanings of the previously overloaded `None` from the rule's perspective. Explicitly, collapsing the meanings and simply having the rule ignore things if the value is `None` is _okay_, but I felt it was a bit better to panic from the rule if the assumption it has is ever broken.

[`serde_with` uses a similar concept which I like quite a lot](https://docs.rs/serde_with/latest/serde_with/rust/double_option).

Though it introduces a small bit of complexity, it also makes things more explicit and provides an additional fail-safe if that assumption is every broken, so I'd argue to keep it (though I could be persuaded not to if the team felt strongly otherwise).

Comment (peterhuene) [reply wdl-lint/src/rules/deprecated_placeholder_option.rs:69]: I think by the nature of the CST having a root node and the root node being the "document", this invariant is very unlikely to ever be violated unless something is catastrophically wrong.

If we change this to `Option<SupportedVersion>`, then if somehow we violate the invariant, the rule won't fire a diagnostic rather than panicking like it does now with the double option; either way we'll observe the problem in our test coverage.

To me it just complicates the code (albeit it only slightly) and I certainly do a double take when I see a double option.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [reply wdl-lint/src/rules/runtime_section_keys.rs:360]: There is a distinction between assigning `Default::default()` and calling `clear()`: the latter will preserve the capacity. 

While it's a minor thing, I think reducing allocations when it's trivial to do so would be ideal. That said, I don't think it's worth holding up the PR unless you'd like to fix.

File: wdl-lint/src/rules/deprecated_placeholder_option.rs
Comment (claymcleod) [reply wdl-lint/src/rules/deprecated_placeholder_option.rs]: Okay, I've finally found some time to dig into the details of the packages and how they work together, and now I have a better idea of the inner mechanisms here. But first, since this may come up as we continue to jointly develop out the package as a group, I should briefly share my perspective on the first part of your comment.

> I think by the nature of the CST having a root node and the root node being the "document", this invariant is very unlikely to ever be violated unless something is catastrophically wrong.
>
> If we change this to Option<SupportedVersion> and somehow we violate the invariant, the rule won't fire a diagnostic rather than panicking like it does now with the double option; either way we'll observe the problem in our test coverage.

To my eye, the difference in approach boils down to whether the benefit of being explicit (both in terms of the code as well as the error messages themselves) is worth the cost of introducing complexity or ugliness into the code.

As you say, it is quite unlikely that the assumption you laid out will ever be broken. However, in the case where it _is_ broken, I will always prefer to immediately get back an error telling me exactly what assumption was violated rather spending time tracking down why the majority of the tests are suddenly failing. Further, when there are two different reasons a `version` could be missing (in this case, because a catastrophic failure occurred in the parsing versus whether the version exists but failed to be parsed to a supporting version), I always prefer to make that immediately obvious and wrestle with those cases out in the open: the practical implications may be the same such that they could be collapsed, but the context with which I collapsed them is far more likely to be missed the next time I visit that code. You could encode those assumptions in comments, but I prefer for the code to yell at me when it can (😄).

As you might intuit from this perspective, I'm quite a defensive software engineer, as I've been bitten in big ways more than once by not taking this approach. However, I'm happy to be reigned in and accept the risks associated with not being explicit, as the likelihood that either approach will be consequential in the long term is small.

---

> Alternatively, we could also unwrap a Option<SupportedVersion> and continue to panic as visitors should not be receiving this callback on an unsupported document version (we skip parsing the contents of the file).

Now that I have a better understanding of what's going on underneath the hood, this comment spurred me to think that there is a better solution hidden behind a more fundamental question: do we want to allow `Visitor`s to traverse documents that aren't a supported WDL version?

To illustrate what happens today, take this document:

```
version 2.0

task {}
workflow {}
```

The root node and preamble are parsed, and the rest of the document is left as an unparsed node:

```
RootNode@0..33
  VersionStatementNode@0..11
    VersionKeyword@0..7 "version"
    Whitespace@7..8 " "
    Version@8..11 "2.0"
  Unparsed@11..33 "\n\ntask {}\nworkflow {}\n"
```

On its face, there is no reason not to allow this AST to be visited by the validator: documents with unsupported WDL versions produce an AST just the same as any supported versions. The big difference is that these ASTs only contain `RootNode` and the various version nodes, so only the `document()` and `version_statement()` callbacks will be executed.

That said, after giving it some deeper thought, I think there is very little benefit to allowing this behavior.

- I think there are very few pros to be had. Basically nothing can be confidently asserted about unsupported versions besides the fact that they are unsupported—any rules that do manage to execute could produce false positive or false negative results in a "silent" manner. Practically speaking, no callbacks besides `document()` and `version_statement_node()` would be run on this kind of AST, so none of the current rules will execute properly.
- The cost of allowing it means that every rule needs to now consider a greater surface area of whether the rule is valid for non-supported versions (a strange concept in general to me—the versions are unsupported by definition). This will lead to boilerplate that is needed in any version-dependent rule, which will become tiring.

Today, we almost always manually check if the parser produced any diagnostics before moving forward with validation/linting. For example, in the `wdl lint` command:

```
 async fn exec(self) -> Result<()> {
        let source = read_source(&self.path)?;
        let (document, diagnostics) = Document::parse(&source);

        // HERE, WE CHECK TO MAKE SURE NO DIAGNOSTICS OCCURRED WHILE PARSING
        // AND RETURN IF ANY ARE ENCOUNTERED.
        if !diagnostics.is_empty() {
            emit_diagnostics(&self.path.to_string_lossy(), &source, &diagnostics)?;

            bail!(
                "aborting due to previous {count} diagnostic{s}",
                count = diagnostics.len(),
                s = if diagnostics.len() == 1 { "" } else { "s" }
            );
        }

      
        let mut validator = Validator::default();
        validator.add_visitor(LintVisitor::default());
        if let Err(diagnostics) = validator.validate(&document) {
            emit_diagnostics(&self.path.to_string_lossy(), &source, &diagnostics)?;

            bail!(
                "aborting due to previous {count} diagnostic{s}",
                count = diagnostics.len(),
                s = if diagnostics.len() == 1 { "" } else { "s" }
            );
        }

        Ok(())
    }
```

There's only one place I spotted in the current implementation where we don't take this approach, and it's in [the engine code](https://github.com/stjude-rust-labs/wdl/blob/main/wdl-analysis/src/engine.rs#L552). When I was designing a solution that followed @peterhuene's suggestion to `unwrap()` the version, this was the cause of inconsistent behavior depending on what happened when I used `wdl lint` to test my solution versus when the engine took over during unit tests.

I'm still open to saying that we want to take the philosophy of allowing unsupported WDL documents to be visited, but, if we do that, we should probably go the full mile and implemented the `supported(version: Option<SupportedVersion>) -> bool` function to make this dance a little more graceful.

Comment (peterhuene) [reply wdl-lint/src/rules/deprecated_placeholder_option.rs]: Agreed, I think having `supported` on the rules / visitors makes a lot of sense.

I'm in support of resolving this issue as you see fit as I know completely understand why it's a double option 👍 .

Regarding the way that analysis is behaving, we could certainly make that consistent with everything else and only extend the diagnostics for a validation visit if parsing didn't have any documents.

Comment (claymcleod) [reply wdl-lint/src/rules/deprecated_placeholder_option.rs]: My preference is to roll with exactly what is implemented right now and merge it in so it makes the Sprocket Docket on Monday. My gut is that the `supported` callback is a bigger change, and this PR accomplishes enough such that we should kick that can down the road.

@a-frantz and @adthrasher, I plan to merge this Monday if there are no other objections, as Peter and I appear to land in the same place on this.

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (claymcleod) [reply wdl-lint/src/rules/runtime_section_keys.rs]: Good catch here, I corrected this.

**Review by peterhuene — APPROVED** [review]

File: wdl-lint/src/rules/runtime_section_keys.rs
Comment (peterhuene) [reply wdl-lint/src/rules/runtime_section_keys.rs:297]: Accordingly, the only value without a default is `container` (`docker`):

> The user is strongly suggested to specify a container for every task. There is no default value for container. If container is not specified, the execution behavior is determined by the execution engine. Typically, the task is simply executed in the host environment.

**Review by a-frantz — APPROVED** [review]

**Pull request merged**
